File: python/src/static_memory_index.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

// 本檔案實作了 `StaticMemoryIndex` 類別的 Python 封裝。
// 這個類別允許 Python 載入一個預先建立好的靜態記憶體索引，並對其執行搜尋操作。

#include "static_memory_index.h"

#include "pybind11/numpy.h"

namespace diskannpy
{

// 輔助函式，用於建立一個為載入靜態索引而設定的 `diskann::Index` 物件。
// 它會設定 `dynamic_index = false` 和 `enable_tags = false` 等參數。
template <class DT>
diskann::Index<DT, StaticIdType, filterT> static_index_builder(const diskann::Metric m, const size_t num_points,
                                                               const size_t dimensions,
                                                               const uint32_t initial_search_complexity)
{
    if (initial_search_complexity == 0)
    {
        throw std::runtime_error("initial_search_complexity must be a positive uint32_t");
    }
    auto index_search_params = diskann::IndexSearchParams(initial_search_complexity, omp_get_num_procs());
    return diskann::Index<DT>(m, dimensions, num_points,
                              nullptr,                                                           // 索引寫入參數 (載入時不需要)
                              std::make_shared<diskann::IndexSearchParams>(index_search_params), // 索引搜尋參數
                              0,       // 凍結點數量
                              false,   // 非動態索引
                              false,   // 不啟用標籤
                              false,   // 無並行整理
                              false,   // 不使用 PQ 建立
                              0,       // PQ 區塊數
                              false);  // 不使用 OPQ
}

// StaticMemoryIndex 的建構函式 (對應 Python 的 __init__)
template <class DT>
StaticMemoryIndex<DT>::StaticMemoryIndex(const diskann::Metric m, const std::string &index_prefix,
                                         const size_t num_points, const size_t dimensions, const uint32_t num_threads,
                                         const uint32_t initial_search_complexity)
    : _index(static_index_builder<DT>(m, num_points, dimensions, initial_search_complexity))
{
    const uint32_t _num_threads = num_threads != 0 ? num_threads : omp_get_num_procs();
    // 在物件初始化時，立即呼叫底層 C++ Index 物件的 load 方法來載入索引檔案。
    _index.load(index_prefix.c_str(), _num_threads, initial_search_complexity);
}

// 單一查詢的搜尋方法
template <typename DT>
NeighborsAndDistances<StaticIdType> StaticMemoryIndex<DT>::search(
    // pybind11 會自動將傳入的 NumPy 陣列轉換為 py::array_t，無需手動轉換。
    py::array_t<DT, py::array::c_style | py::array::forcecast> &query, const uint64_t knn, const uint64_t complexity)
{
    // 建立用於儲存結果的 NumPy 陣列
    py::array_t<StaticIdType> ids(knn);
    py::array_t<float> dists(knn);
    std::vector<DT *> empty_vector;
    _index.search(query.data(), knn, complexity, ids.mutable_data(), dists.mutable_data());
    
    // 返回一個包含 ID 和距離的元組，pybind11 會將其轉換為 Python 的 tuple of numpy arrays。
    return std::make_pair(ids, dists);
}

// 帶過濾條件的單一查詢搜尋方法
template <typename DT>
NeighborsAndDistances<StaticIdType> StaticMemoryIndex<DT>::search_with_filter(
    py::array_t<DT, py::array::c_style | py::array::forcecast> &query, const uint64_t knn, const uint64_t complexity,
    const filterT filter)
{
    py::array_t<StaticIdType> ids(knn);
    py::array_t<float> dists(knn);
    std::vector<DT *> empty_vector;
    _index.search_with_filters(query.data(), filter, knn, complexity, ids.mutable_data(), dists.mutable_data());
    return std::make_pair(ids, dists);
}

// 批次搜尋方法
template <typename DT>
NeighborsAndDistances<StaticIdType> StaticMemoryIndex<DT>::batch_search(
    py::array_t<DT, py::array::c_style | py::array::forcecast> &queries, const uint64_t num_queries, const uint64_t knn,
    const uint64_t complexity, const uint32_t num_threads)
{
    const uint32_t _num_threads = num_threads != 0 ? num_threads : omp_get_num_procs();
    // 建立二維的 NumPy 陣列以儲存批次搜尋的結果
    py::array_t<StaticIdType> ids({num_queries, knn});
    py::array_t<float> dists({num_queries, knn});
    std::vector<DT *> empty_vector;

    omp_set_num_threads(static_cast<int32_t>(_num_threads));

    // 使用 OpenMP 將查詢分配到多個執行緒上平行執行，以提高 QPS。
#pragma omp parallel for schedule(dynamic, 1) default(none) shared(num_queries, queries, knn, complexity, ids, dists)
    for (int64_t i = 0; i < (int64_t)num_queries; i++)
    {
        // `queries.data(i)` 和 `ids.mutable_data(i)` 用於存取 NumPy 陣列的第 i 行。
        _index.search(queries.data(i), knn, complexity, ids.mutable_data(i), dists.mutable_data(i));
    }

    return std::make_pair(ids, dists);
}

// 模板實例化
template class StaticMemoryIndex<float>;
template class StaticMemoryIndex<uint8_t>;
template class StaticMemoryIndex<int8_t>;

} // namespace diskannpy

File: python/src/_static_memory_index.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT license.

# 本檔案定義了 `StaticMemoryIndex` 類別，這是使用者與靜態記憶體索引互動的主要 Python 介面。
# 它封裝了底層的 C++ 綁定，提供了參數驗證和更易於使用的 API。

import json
import os
import warnings
from typing import Optional

import numpy as np

from . import (
    DistanceMetric,
    QueryResponse,
    QueryResponseBatch,
    VectorDType,
    VectorLike,
    VectorLikeBatch,
)
from . import _diskannpy as _native_dap
from ._common import (
    _assert,
    _assert_is_nonnegative_uint32,
    _assert_is_positive_uint32,
    _castable_dtype_or_raise,
    _ensure_index_metadata,
    _valid_index_prefix,
    _valid_metric,
)

__ALL__ = ["StaticMemoryIndex"]


class StaticMemoryIndex:
    """
    一個不可變的、完全載入到記憶體中的 DiskANN 索引。
    """

    def __init__(
        self,
        index_directory: str,
        num_threads: int,
        initial_search_complexity: int,
        index_prefix: str = "ann",
        distance_metric: Optional[DistanceMetric] = None,
        vector_dtype: Optional[VectorDType] = None,
        dimensions: Optional[int] = None,
        enable_filters: bool = False
    ):
        """
        建構函式，用於載入一個已存在的靜態記憶體索引。

        ### 參數
        - **index_directory**: 包含索引檔案的目錄。
        - **num_threads**: 搜尋時使用的執行緒數。
        - **initial_search_complexity**: 初始化時為搜尋分配的候選集大小。
        - ... (其他參數)
        """
        index_prefix_path = _valid_index_prefix(index_directory, index_prefix)
        self._labels_map = {}
        self._labels_metadata = {}
        if enable_filters:
            try:
                with open(f"{index_prefix_path}_labels_map.txt", "r") as labels_map_if:
                    for line in labels_map_if:
                        (key, val) = line.split("\t")
                        self._labels_map[key] = int(val)
                with open(f"{index_prefix_path}_label_metadata.json", "r") as labels_metadata_if:
                    self._labels_metadata = json.load(labels_metadata_if)
            except: # noqa: E722
                # exceptions are basically presumed to be either file not found or file not formatted correctly
                raise RuntimeException("Filter labels file was unable to be processed.")
        vector_dtype, metric, num_points, dims = _ensure_index_metadata(
            index_prefix_path,
            vector_dtype,
            distance_metric,
            1,  # 在此上下文中 max_vectors 不重要
            dimensions,
        )
        dap_metric = _valid_metric(metric)

        # --- 參數驗證 ---
        _assert_is_nonnegative_uint32(num_threads, "num_threads")
        _assert_is_positive_uint32(
            initial_search_complexity, "initial_search_complexity"
        )

        self._vector_dtype = vector_dtype
        self._dimensions = dims

        # --- 根據資料類型，選擇對應的底層 C++ 原生類別 ---
        if vector_dtype == np.uint8:
            _index = _native_dap.StaticMemoryUInt8Index
        elif vector_dtype == np.int8:
            _index = _native_dap.StaticMemoryInt8Index
        else:
            _index = _native_dap.StaticMemoryFloatIndex

        # --- 實例化並載入 C++ 索引物件 ---
        # 這裡會呼叫 C++ 封裝層的建構函式，進而觸發底層 C++ 索引的 `load` 方法。
        self._index = _index(
            distance_metric=dap_metric,
            num_points=num_points,
            dimensions=dims,
            index_path=index_prefix_path,
            num_threads=num_threads,
            initial_search_complexity=initial_search_complexity,
        )

    def search(
            self, query: VectorLike, k_neighbors: int, complexity: int, filter_label: str = ""
    ) -> QueryResponse:
        """
        對單一查詢向量執行搜尋。

        ### 參數
        - **query**: 一維 NumPy 陣列，其維度和 dtype 必須與索引匹配。
        - **k_neighbors**: 要返回的鄰居數量。
        - **complexity**: 搜尋時使用的候選集大小 (L)。L 越大，召回率越高，但延遲也越高。
        - **filter_label**: (可選) 用於過濾搜尋的標籤字串。
        """
        # --- 搜尋前的參數驗證 ---
        if filter_label != "":
            if len(self._labels_map) == 0:
                raise ValueError(
                    f"A filter label of {filter_label} was provided, but this class was not initialized with filters "
                    "enabled, e.g. StaticDiskMemory(..., enable_filters=True)"
                )
            if filter_label not in self._labels_map:
                raise ValueError(
                    f"A filter label of {filter_label} was provided, but the external(str)->internal(np.uint32) labels map "
                    f"does not include that label."
                )
            k_neighbors = min(k_neighbors, self._labels_metadata[filter_label])
        _query = _castable_dtype_or_raise(query, expected=self._vector_dtype)
        _assert(len(_query.shape) == 1, "query vector must be 1-d")
        _assert(
            _query.shape[0] == self._dimensions,
            f"query vector must have the same dimensionality as the index; index dimensionality: {self._dimensions}, "
            f"query dimensionality: {_query.shape[0]}",
            )
        _assert_is_positive_uint32(k_neighbors, "k_neighbors")
        _assert_is_nonnegative_uint32(complexity, "complexity")

        if k_neighbors > complexity:
            warnings.warn(
                f"k_neighbors={k_neighbors} asked for, but list_size={complexity} was smaller. Increasing {complexity} to {k_neighbors}"
            )
            complexity = k_neighbors

        # --- 呼叫底層 C++ 搜尋函式 ---
        if filter_label == "":
            neighbors, distances = self._index.search(query=_query, knn=k_neighbors, complexity=complexity)
        else:
            filter = self._labels_map[filter_label]
            neighbors, distances = self._index.search_with_filter(
                query=query,
                knn=k_neighbors,
                complexity=complexity,
                filter=filter
            )
        
        # --- 將結果封裝成更易於使用的 NamedTuple 並返回 ---
        return QueryResponse(identifiers=neighbors, distances=distances)


    def batch_search(
        self,
        queries: VectorLikeBatch,
        k_neighbors: int,
        complexity: int,
        num_threads: int,
    ) -> QueryResponseBatch:
        """
        對一批查詢向量執行批次搜尋。這種方式比逐一搜尋效率更高，因為它在底層利用了多執行緒平行處理。

        ### 參數
        - **queries**: 二維 NumPy 陣列，(查詢數量, 維度)。
        - ... (其他參數與 search 類似)
        """

        # --- 批次搜尋前的參數驗證 ---
        _queries = _castable_dtype_or_raise(queries, expected=self._vector_dtype)
        _assert(len(_queries.shape) == 2, "queries must must be 2-d np array")
        _assert(
            _queries.shape[1] == self._dimensions,
            f"query vectors must have the same dimensionality as the index; index dimensionality: {self._dimensions}, "
            f"query dimensionality: {_queries.shape[1]}",
        )
        _assert_is_positive_uint32(k_neighbors, "k_neighbors")
        _assert_is_positive_uint32(complexity, "complexity")
        _assert_is_nonnegative_uint32(num_threads, "num_threads")

        if k_neighbors > complexity:
            warnings.warn(
                f"k_neighbors={k_neighbors} asked for, but list_size={complexity} was smaller. Increasing {complexity} to {k_neighbors}"
            )
            complexity = k_neighbors

        num_queries, dim = _queries.shape
        neighbors, distances = self._index.batch_search(
            queries=_queries,
            num_queries=num_queries,
            knn=k_neighbors,
            complexity=complexity,
            num_threads=num_threads,
        )
        
        # --- 將結果封裝成 NamedTuple 並返回 ---
        return QueryResponseBatch(identifiers=neighbors, distances=distances)
